### PR TITLE
add ex36 tutorial

### PR DESCRIPTION
**Pull request by thatguynoe**
This PR provides a tutorial to ex36, and guides users on how to...
* implement the [Proximal-Garlekin](https://arxiv.org/abs/2307.12444) method in MFEM,
* construct discretizations of nonlinear problems, and
* solve nonlinear problems in MFEM.

### CODE REVIEW

**Comment by thatguynoe** [comment]
I'm not sure why my [level four subheadings](https://github.com/mfem/web/blob/235fa34fddc20125b2e4ef1b235ba58ee9428bf4/src/tutorial/obstacle.md?plain=1#L112) subheadings are not showing up in the table of contents.

File: src/tutorial/obstacle.md [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+$$
+K = \\{v \in H^1(\Omega) : v|_{\partial \Omega} = g \text{ and } v \geq \varphi\\},
+$$
+
+where $\Omega$ is an open bounded domain with smooth boundary and $\varphi$ is
+a smooth function representing the obstacle.
+
+We will solve a discretization of this problem using the proximal Galerkin
+finite element method, introduced by [Keith and
+Surowiec](https://arxiv.org/abs/2307.12444). Indeed, given linear subspaces
```
Comment (bkhara) [inline src/tutorial/obstacle.md]: We will discretize and solve this problem using the proximal Galerkin
finite element method, introduced by [Keith and
Surowiec](https://arxiv.org/abs/2307.12444). In this method, given linear subspaces ...

File: src/tutorial/obstacle.md [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+\begin{aligned}
+(\alpha_{k+1} \nabla u_h, \nabla v) + (\psi_h, v) &= (\alpha_{k+1} f + \psi_h^k, v) && \text{for all } v \in V_h, \\\\
+(u_h,w) - (\exp \psi_h, w) &= (\varphi,w) && \text{for all } w \in W_h,
+\end{aligned}
+\end{equation}
+$$
+
+where $\alpha_k > 0$ is a sequence of step sizes and $g_h \in H^1(\Omega)$
+provides an approximation of the boundary values $g_h|_{\partial \Omega}
+\approx g$. In order to accomplish this, we apply the Quasi-Newton method[^1] and will solve the following linearized discrete saddle-point problem: find $u_h \in V_h$ and $\delta \psi_h \in W_h$ such that
```
[the remark this answers is not in the document]
Comment (bkhara) [reply src/tutorial/obstacle.md]: please add a step here showing why you are solving for for u_h instead of delta u_h)

File: src/tutorial/obstacle.md [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+
+Below we highlight selected portions of the example code and connect them with
+the description in the previous section. You can follow along by browsing
+`ex36.cpp` in your VS Code browser window. In the settings of this tutorial,
+the visualization will automatically update in the GLVis browser window.
+
+#### The Mesh
+
+The computational mesh for this problem is set to `disc-nurbs.mesh` in line
+[77](https://github.com/mfem/mfem/blob/master/examples/ex36.cpp#L109). We can
```
Comment (bkhara) [inline src/tutorial/obstacle.md]: > [77]

[109]

File: src/tutorial/obstacle.md [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+GridFunction psi_old_gf(&L2fes);
+GridFunction psi_gf(&L2fes);
+```
+
+Next, we define the function coefficients for the solution and use them to
+initialize the initial guess. We also define the exact solution of our problem
+with the function `exact_solution_obstacle` defined in lines
+[423-439](https://github.com/mfem/mfem/blob/master/examples/ex36.cpp#L423-L439),
+as well as its gradient `exact_solution_gradient_obstacle` in lines
+[441-459](https://github.com/mfem/mfem/blob/master/examples/ex36.cpp#L441-L459),
```
Comment (bkhara) [inline src/tutorial/obstacle.md]: [441-459](https://github.com/mfem/mfem/blob/master/examples/ex36.cpp#L441-L459). We will need these for computing the error in our finite element solution. (Please add a line here saying that we need our functions in the form of FunctionCoefficent and VectorFunctionCoefficient (when applicable)) Recall that the function `spherical_obstacle` represents our
obstacle function $\varphi$ in (4).

File: src/tutorial/obstacle.md [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+psi_old_gf = psi_gf;
+```
+
+#### Solving the System
+
+The main iteration loop is found in lines
+[211-343](https://github.com/mfem/mfem/blob/master/examples/ex36.cpp#L211-L343),
+and consists of an inner loop in lines
+[223-323](https://github.com/mfem/mfem/blob/master/examples/ex36.cpp#L223-L323)
+which implements (4), the Quasi-Newton method mentioned earlier.
```
Comment (bkhara) [inline src/tutorial/obstacle.md]: Newton's method

File: src/tutorial/obstacle.md [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+```c++
+b0.AddDomainIntegrator(new DomainLFIntegrator(alpha_f));
+b0.AddDomainIntegrator(new DomainLFIntegrator(psi_old_minus_psi));
+b0.Assemble();
+```
+
+The linear system $AX = B$ is solved using [GMRES](https://en.wikipedia.org/wiki/Generalized_minimal_residual_method), using a block diagonal preconditioner `prec`. In line [297](https://github.com/mfem/mfem/blob/master/examples/ex36.cpp#L297), the matrix $A$ corresponds to `A`, the vector $B$ corresponds to `rhs`, and the solution $X$ is stored in `x`.
+
+```c++
+GMRES(A,prec,rhs,x,0,10000,500,1e-12,0.0);
```
Comment (bkhara) [inline src/tutorial/obstacle.md]: not essential, but maybe consider explaining the important parameters in the GMRES function call

File: src/tutorial/obstacle.md [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+changing the definition of `ref_levels`.
+
+```c++
+for (int l = 0; l < ref_levels; l++)
+{
+  mesh.UniformRefinement();
+}
+```
+
+Next, we control potential geometry error by setting the curvature of our mesh
```
[the remark this answers is not in the document]
Comment (thatguynoe) [reply src/tutorial/obstacle.md]: Addressed in https://github.com/mfem/web/pull/305/commits/98fed6a1f59d79c9ec2b59aaf4703eccffeecb66.

**Comment by brendankeith** [comment]
Hopefully by tomorrow 🤞

**Comment by brendankeith** [comment]
This is now ready for review!

**Comment by tzanio** [comment]
@mlstowell and @acfisher -- can you please review?

File: src/annotated/ex2.md [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+$$
+
+The boundary of the domain, $\partial\Omega$, is divided into two parts, the displacement (Dirichlet) boundary $\Gamma_D$ and the traction (Neumann) boundary $\Gamma_T$. In this example problem, the left-hand side of the boundary has a homogenous ($u = 0$) displacement boundary condition, whereas the rest of the boundary has a traction boundary condition, ${\sigma} \cdot {n} = t$. In this problem, the traction force $t$ is non-zero only on the right-hand side of the boundary. Now using our constitutive law (2), which gives the relation between ${\sigma}$ and the displacement field ${u}$, we obtain the weak formulation
+
+$$
+\int_{\Omega} \lambda (\nabla\cdot u_h)(\nabla\cdot v_h) + \mu (\nabla u_h + \nabla {u}^T_h) : \nabla v_h \, \mathrm{d}{x} = \int_{\Gamma_T} v_h \cdot t\, \mathrm{d}S
+.
+$$
+
+Letting $e_i \in \mathbb{R}^n$ denote the standard Cartesian unit vectors and setting $v_h = \varphi_j e_i$, we can leverage the basis expansion formula (4) to rewrite (6) as
```
Comment (mlstowell) [inline src/annotated/ex2.md]: It might be more accurate to say $v_{i,h} = \varphi_j e_i$.

File: src/annotated/ex2.md [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+The source file [examples/ex2.cpp](https://github.com/mfem/mfem/blob/master/examples/ex2.cpp) implements the above FEM for a beam-shaped body consisting of two adjacent material subdomains.
+In particular, it sets $t=-0.01$ on the right-hand side of the boundary and enforces the Dirichlet and Neumann boundary conditions over the other corresponding subsets of the boundary.
+
+Below we highlight selected portions of the example code and connect them with
+the description in the previous section. You can follow along by browsing
+`ex2.cpp` in your VS Code browser window. In the settings of this tutorial, the
+visualization will automatically update in the GLVis browser window.
+
+MFEM uses `attributes` and `bdr_attributes` to distinguish between subdomains and subsets of the domain boundary, respectively.
+The computational mesh is provided as input (option `-m`) that could be 2D or 3D domain made up of triangular/quadrilateral/tetrahedral/hexahedral elements, etc. (It defaults to `beam-tri.mesh` in line
```
Comment (mlstowell) [inline src/annotated/ex2.md]: Missing article: "...could be a 2D or 3D domain..."

File: src/annotated/ex2.md [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+[50](https://github.com/mfem/mfem/blob/master/examples/ex2.cpp#L50).) The code in
+lines [73-84](https://github.com/mfem/mfem/blob/master/examples/ex2.cpp#L73-L84)
+loads the mesh from the given file, `mesh_file` and creates the corresponding MFEM
+pointer `*mesh` of class `Mesh`. It then checks that the mesh file uploaded has at least two different materials and correct number of boundary attributes.
```
Comment (mlstowell) [inline src/annotated/ex2.md]: Missing article: "...and the correct number..."

File: src/annotated/ex6.md [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+$$
+-\Delta u = f,
+$$
+
+where _u_ is the potential field and _f_ is the source function. This PDE is a
+generalization of the [_Laplace Equation_](https://en.wikipedia.org/wiki/Laplace%27s_equation).
+
+To approximately solve the above continuous equation on computers, we need to
+[discretize](https://en.wikipedia.org/wiki/Discretization) it by introducing a
+finite (discrete) number of unknowns to compute for. In the
```
Comment (mlstowell) [inline src/annotated/ex6.md]: "for" is not really necessary here

File: src/annotated/ex6.md [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+
+### <i class="fa fa-check-square-o"></i>&nbsp; Annotated Example 6
+
+Below is the classification of each mesh file used in the sample runs, indicating the dimension, element type, and whether it is conforming or non-conforming:
+
+| Mesh File                   | Dimension   | Element Type                 | Classification    |
+|-----------------------------|-------------|-------------------------------|-------------------|
+| `square-disc.mesh`          | 2D          | triangles                     | conforming        |
+| `square-disc-nurbs.mesh`    | 2D          | NURBS quadrilaterals          | non-conforming    |
+| `star.mesh`                 | 2D          | triangles                     | conforming        |
```
Comment (mlstowell) [inline src/annotated/ex6.md]: star.mesh is a quadrilateral mesh

File: src/annotated/ex6.md [hunk 13]
```diff
@@ -0,0 +1,485 @@
...
+| `escher.mesh`               | 2D          | quadrilaterals                | non-conforming    |
```
Comment (mlstowell) [inline src/annotated/ex6.md]: escher.mesh is a 3D tetrahedral conforming mesh

File: src/annotated/ex6.md [hunk 14]
```diff
@@ -0,0 +1,485 @@
...
+      mesh.UniformRefinement();
+   }
+   mesh.SetCurvature(2);
+}
+```
+
+---
+
+#### Defining Variables and Spaces
+In the next block of code, we create the finite element space, i.e., specify the finite element basis functions on the mesh. This involves the MFEM class and `FiniteElementSpace`, which connects the space and the mesh. The code in [lines 112-113](https://github.com/mfem/mfem/blob/master/examples/ex6.cpp#L112-113) is essentially:
```
Comment (mlstowell) [inline src/annotated/ex6.md]: Did you mesh "MFEM class, `FiniteElementSpace`, which connects..."? Otherwise I'm not sure which "MFEM class" you are referring to.

File: src/annotated/ex2.md [hunk 15]
```diff
@@ -0,0 +1,430 @@
...
+else
+{
+    FiniteElementCollection *fec = new H1_FECollection(order, dim);
+    FiniteElementSpace *fespace = new FiniteElementSpace(mesh, fec, dim);
+}
+cout << "Number of finite element unknowns: " << fespace->GetTrueVSize() << endl << "Assembling: " << flush;
+```
+
+The printed number of finite element unknowns corresponds to the
+size of the linear system $mn$ from the previous section.
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply src/annotated/ex2.md]: The solution variable has n components, while the scalar basis has m elements, leading to a linear system of size mn

File: src/annotated/ex3.md [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+
+---
+[_Maxwell's equations_](https://en.wikipedia.org/wiki/Maxwell%27s_equations) are a set of coupled PDEs that form the basis for the study of electromagnetism. This annotated example demonstrates how to solve the second-order definite Maxwell equation,
+$$
+\nabla \times (\mu^{-1} \nabla \times E) + \sigma E = f,
+$$
+corresponding to a simple model of electromagnetic diffusion with parameters $\mu$, $\sigma>0$. The boundary condition is given by setting $E \times n$ equal to some given tangential field $g$. 
+
+After taking the inner product of (1) with a test vector field $F$, followed by integrating over the domain $\Omega$, we can apply [Stokes' theorem](https://en.wikipedia.org/wiki/Stokes_theorem) to arrive at the following variational formulation: find $E \in H(\text{curl})$ with $E \times n = g$ on the boundary of our domain $\partial \Omega$ such that 
+ $$a(E,F) := (\mu^{-1} \nabla \times E, \nabla \times F) + (\sigma E, F) = (f,F)=: b(F)$$
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply src/annotated/ex3.md]: Maybe I'm confused, but the equation is centered when I render it.

Image

File: src/annotated/ex3.md [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+Review  <a href="../../tutorial/fem"><i class="fa fa-play-circle"></i>&nbsp; Examples 1</a> and <a href="../ex2"><i class="fa fa-play-circle"></i>&nbsp; 2</a> before starting this lesson.
+</div>
+</div>
+
+---
+[_Maxwell's equations_](https://en.wikipedia.org/wiki/Maxwell%27s_equations) are a set of coupled PDEs that form the basis for the study of electromagnetism. This annotated example demonstrates how to solve the second-order definite Maxwell equation,
+$$
+\nabla \times (\mu^{-1} \nabla \times E) + \sigma E = f,
+$$
+corresponding to a simple model of electromagnetic diffusion with parameters $\mu$, $\sigma>0$. The boundary condition is given by setting $E \times n$ equal to some given tangential field $g$.
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply src/annotated/ex3.md]: Good point

File: src/annotated/ex3.md [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+
+---
+
+#### Mesh creation
+```cpp
+   Mesh *mesh = new Mesh(mesh_file, 1, 1);
+   dim = mesh->Dimension();
+   int sdim = mesh->SpaceDimension();
+```
+This block of code defines the mesh after reading from the `mesh_file` provided by the user. The `1, 1` arguments typically relate to generating boundary elements and refining uniform features, respectively. Then `dim` retrieves the dimension of the mesh elements (e.g., 2 for triangles, 3 for tetrahedrons). Then `sdim` retrieves the dimension of the ambient space.
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply src/annotated/ex3.md]: Much better :)

File: src/annotated/ex3.md [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+#### Finite element space definition
+The next block of code defines the Finite Element Space on the mesh provided by the user.
+
+```cpp
+   FiniteElementCollection *fec = new ND_FECollection(order, dim);
+   FiniteElementSpace *fespace = new FiniteElementSpace(mesh, fec);
+   cout << "Number of finite element unknowns: "
+        << fespace->GetTrueVSize() << endl;
+```
+Here, the `FiniteElementCollection` corresponds to Nédélec basis functions because Nédélec finite elements have optimal approximation approximation properties in $H(\mathrm{curl})$; i.e., the energy space over which problem (2) is well-defined.
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply src/annotated/ex3.md]: I added the following remark:
```The Nédélec basis functions are intrinsically vector-valued. In particular, they cannot be constructed as a Cartesian product of a scalar basis, unlike the vector-valued basis used in [Example 2](http://127.0.0.1:8000/annotated/ex2) for linear elasticity. In turn, these basis functions are typically nonzero in all components, and their degree-of-freedom numbering does not represent their active vector components.```

File: src/annotated/ex6.md [hunk 20]
```diff
@@ -0,0 +1,485 @@
...
+The method `FormLinearSystem` takes the bilinear form `a`, linear form `b`, current solution `x`, and a list of essential boundary degrees of freedom `ess_tdof_list`. It eliminates boundary conditions, handles any constraints, and sets up the system for the true degrees of freedom only. The flag `copy_interior = 1` ensures that the interior (unconstrained) DOFs are copied properly for consistency.
+
+```c++
+OperatorPtr A;
+Vector B, X;
+const int copy_interior = 1;
+a.FormLinearSystem(ess_tdof_list, x, b, A, X, B, copy_interior);
+```
+
+This process produces `A` (the assembled system matrix), `B` (the right-hand side vector), `X` (the unknowns vector to solve for).
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply src/annotated/ex6.md]: Replaced "(the assembled system matrix)" with "(an `OperatorPtr` for the assembled system matrix)."

**Comment by brendankeith** [comment]
Thanks for your comments, @mlstowell and @acfisher!

Please let me know if I missed anything!
